### ISSUE TITLE
Optimize Session Startup by Parallelizing Component Initialization

### ISSUE THREAD

**Issue by golbin** [pipecat-ai/pipecat#904]
# Description
Is this reporting a bug or feature request?  
This is a feature request to optimize the session startup process.

### Issue description
The current session startup process initializes components (TTS, LLM, STT, and Daily Join) sequentially, which results in higher latency before the session becomes operational. By refactoring the process to initialize these components in parallel, the startup time can be significantly reduced.

### Expected behavior
The initialization of TTS, LLM, STT, and Daily Join should occur in parallel, resulting in a noticeable reduction in session startup time.

### Actual behavior
Components are initialized sequentially, leading to a slower startup time.

**Comment by aconchillo** [comment]
Thank you for reporting this @golbin . The reason behind the current behavior is to make all the components are successfully initialzed in order so we don't push frames prematurely. But yes, maybe we could possibly initialize everything in parallel and just wait for the last service being initialized before starting pushing frames.

**Comment by golbin** [comment]
Yes, handling parallel processing might be a bit challenging, but it could significantly reduce the startup time.

If we can reuse certain components, the startup could become extremely fast. Additionally, if we can maintain connections to TTS and STT even after the session ends, we could potentially onboard new users with almost no waiting time.

For a simpler approach, we could start by parallelizing only the Daily transport. This alone could substantially reduce the waiting time, as the time spent on the Daily join process is roughly equivalent to the combined time required to connect to TTS and STT.

| Time                   | Event                                     | Time Difference |
|------------------------|-------------------------------------------|-----------------|
| 2024-12-21 10:36:27.580 | Modules: Session started                 | 0.001 seconds   |
| 2024-12-21 10:36:27.580 | VAD: Set VAD parameters                  | 0.000 seconds   |
| 2024-12-21 10:36:27.649 | Audio: Silero VAD loaded                 | 0.069 seconds   |
| 2024-12-21 10:36:27.684 | Processors: Link frame processors        | 0.035 seconds   |
| 2024-12-21 10:36:27.909 | Services: Connect to Cartesia            | 0.225 seconds   |
| 2024-12-21 10:36:28.178 | Daily: Join session                      | 0.269 seconds   |
| 2024-12-21 10:36:28.616 | Daily: Joined session                    | 0.438 seconds   |
| 2024-12-21 10:36:29.042 | Modules: Sync session state              | 0.426 seconds   |
| 2024-12-21 10:36:29.043 | Modules: Update session state            | 0.001 seconds   |
| 2024-12-21 10:36:30.132 | Daily: Participant joined                | 1.089 seconds   |
| 2024-12-21 10:36:30.133 | Modules: Participant joined session      | 0.001 seconds   |
| 2024-12-21 10:36:30.334 | Pipeline: Start session with greeting    | 0.201 seconds   |
| 2024-12-21 10:36:30.335 | Cartesia: Generate TTS for greeting      | 0.001 seconds   |
| 2024-12-21 10:36:30.508 | Bot: Started speaking                    | 0.173 seconds   |

**Comment by golbin** [comment]
We could consider using a Notifier or “PipelineReadyFrame” in a way similar to EndFrame. What do you think?

**Comment by golbin** [comment]
Each service could emit an “XXXReadyFrame” when it becomes ready. Each time this happens, the runner could scan the entire pipeline’s status, and if all services are ready, it could trigger the “StartFrame.”

Concurrency issues might arise when checking the status of each service or due to state changes. However, these issues could likely be resolved by ensuring that “StartFrame” verifies the readiness of the entire pipeline before being triggered.

**Comment by golbin** [comment]
I think we can do it now, thanks to several improvements in traveling and controlling frames that make it possible.

**Comment by golbin** [comment]
I made a PR: https://github.com/pipecat-ai/pipecat/pull/2215

**Comment by aconchillo** [comment]
I was just thinking about this. Once we started queuing system frames (like `StartFrame`) in 0.0.78 (August 2025) and since we push `StartFrame` before actually calling any internal `start()` method (see below), I believe we are currently doing things in parallel already. That is:

- `PipelineTask` pushes `StartFrame`.
- Let's say we reach a transport or a service:
```
            # Push StartFrame before start(), because we want StartFrame to be
            # processed by every processor before any other frame is processed.
            await self.push_frame(frame, direction)
            await self.start(frame)
```
- The code above makes it so we first push `StartFrame` and since it's queued we exist `push_frame()` immediately.
- The following services in the pipeline will do the same, that is push StartFrame` first and then execute any code.

So, I believe this was fixed after we queued system frames. You just need to make sure you push `StartFrame` first though.

**Comment by aconchillo** [comment]
Closing. This was fixed after system frames were queued in Pipecat v0.0.78.

**Comment by aconchillo** [comment]
This has now been reworked quite a bit with a different approach. Note that even with system frames being queued we still had an issue with `StartFrame` being pushed after a service was connecting. So, unless we did the following everywhere:

```
if isinstance(frame, StartFrame):
    await self.push_frame(frame)
    await self._connect()
```

it was not possible to start processors concurrently. And more specifically, `AIService` calls `start()` first which is where connections might happens and then pushing `StartFrame` is next breaking the approach above.

 See https://github.com/pipecat-ai/pipecat/pull/5316
